### ISSUE TITLE
WFS issue

### ISSUE THREAD

**Issue by Matt914** [jjimenezshaw/Leaflet.Control.Layers.Tree#34]
`     <!doctype html>
      <html>
    
      <head>
    
     <meta charset="utf-8" />

    
     <meta name="viewport" content="width=device-width, initial-scale=1">
    <title>Webmap Comune di Colleferro</title>

     <link rel="stylesheet" href="https://unpkg.com/leaflet@1.7.1/dist/leaflet.css" integrity="sha512-xodZBNTC5n17Xt2atTPuE1HxjVMSvLVW9ocqUKLsCC5CXdbqCmblAshOMAS6/keqq/sMZMZ19scR4PsZChSR7A==" crossorigin=""/>
     <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
     <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha384-wvfXpqpZZVQGK6TAh5PVlGOfQNHSoD2xbE+QkPxCAFlNEevoEH3Sl0sibVcOQVnN" crossorigin="anonymous">
     <script src="https://unpkg.com/leaflet@1.7.1/dist/leaflet.js" integrity="sha512-XQoYMqMTK8LvdxXYG3nZ448hOEQiglfqkJs1NOQV44cWnUrBc8PkAOcXy20w0vlaXaVUearIOBhiXZ5V3ynxwA==" crossorigin=""></script>
     <script src="https://code.jquery.com/jquery-3.2.1.slim.min.js" integrity="sha384-KJ3o2DKtIkvYIK3UENzmM7KCkRr/rE9/Qpg6aAZGJwFDMVNA/GpGFF93hXpG5KkN" crossorigin="anonymous"></script>
     <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js" integrity="sha384-ApNbgh9B+Y1QKtv3Rn7W3mgPxhU9K/ScQsAP7hUibX39j7fakFPskvXusvfa0b4Q" crossorigin="anonymous"></script>
     <script src="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/js/bootstrap.min.js" integrity="sha384-JZR6Spejh4U02d8jOt6vLEHfe/JQGiRRSQQxSfFWpi1MquVdAyjUar5+76PVCmYl" crossorigin="anonymous"></script>
     <script src="https://ajax.googleapis.com/ajax/libs/jquery/3.5.1/jquery.min.js"></script>
     <script src="https://code.jquery.com/jquery-3.5.1.min.js" integrity="sha256-9/aliU8dGd2tb6OSsuzixeV4y/faTqgFtohetphbbj0=" crossorigin="anonymous"></script>
     <script src="src/plugins/leaflet-bing-layer.js"></script>
     <script src="https://unpkg.com/leaflet.vectorgrid@latest/dist/Leaflet.VectorGrid.bundled.js"></script>
     <script src="src/plugins/leaflet-providers.js"></script>
     <script src="src/plugins/leaflet.ajax.min.js"></script>
     <script src="src/plugins/proj4.js"></script>
     <script src="src/plugins/proj4leaflet.js"></script>
     <script src="src/plugins/leaflet.wms.js"></script>
     <script src="https://cdn.polyfill.io/v2/polyfill.min.js?features=Promise"></script>

     <style type="text/css">
        html,
        body {
            width: 100%;
            height: 100%;
            margin: 0;
       }

       #map {
           width: 100%;
          height: 100%;
       }    
   
       </style>
    </head>

        <body>
            <div id="map">
                </div>
                    <link rel="stylesheet" href="src/css/L.Control.Layers.Tree.css" crossorigin="" />
                        <script src="src/plugins/L.Control.Layers.Tree.js">
                            </script>

    <script type="text/javascript">
     var center = [41.72707235373704, 13.004139599641844];

     //inserimento iniziale Layers

    var BING_KEY = 'AljSdfpCwtR7wFswYzmgqltJa9xeQvVtoYb1bEuV7jVHowADqs76GPBdEVoAxvuP'
    var binglayer = L.tileLayer.bing(BING_KEY)

    //come inserire WMS ((DA PROVARE CON SERVIZIO WFS))

       var osm = L.tileLayer(
        '//{s}.tile.openstreetmap.org/{z}/{x}/{y}.png', {
          attribution: '© OpenStreetMap contributors'
       }
    );

       var thunderAttr = {
        attribution: '© OpenStreetMap contributors. Tiles courtesy of Andy Allan'
       }
      var transport = L.tileLayer(
       '//{s}.tile.thunderforest.com/transport/{z}/{x}/{y}.png',
       thunderAttr
    );

      var cycle = L.tileLayer(
       '//{s}.tile.thunderforest.com/cycle/{z}/{x}/{y}.png',
        thunderAttr
      );

      var stadiadark = L.tileLayer('https://tiles.stadiamaps.com/tiles/alidade_smooth_dark/{z}/{x}/{y}{r}.png', {
        attribution: '&copy; <a href="https://stadiamaps.com/">Stadia Maps</a>, &copy; <a 
        href="https://openmaptiles.org/">OpenMapTiles</a> &copy; <a href="http://openstreetmap.org">OpenStreetMap</a> 
        contributors'
      });

      var boundaries = L.tileLayer.wms('http://localhost:8080/geoserver/colleferro.territorio/wms', {
        layers: 'colleferro.territorio:confini_amministrativi',
        format: 'image/png',
        transparent: true,
       });

      var owsrootUrl = 'http://localhost:8080/geoserver/colleferro.territorio/ows';

      var defaultParameters = {
         service: 'WFS',
         version: '2.0',
        request: 'GetFeature',
        typeName: 'colleferro.territorio:strade_colleferro',
        outputFormat: 'text/javascript',
        format_options: 'callback:getJson',
        SrsName: 'EPSG:4326'
       };

      var parameters = L.Util.extend(defaultParameters);
      var URL = owsrootUrl + L.Util.getParamString(parameters);

     var WFSLayer = L.geoJson(null, {
            style: function(feature) {
            return {
            stroke: true,
            fillColor: '#B04173',
            fillOpacity: 2,
            color: '#000000',
            weight: 1,
          };
        }
     });

    var ajax = $.ajax({
        url: URL,
       dataType: 'jsonp',
       jsonpCallback: 'getJson',
       success: function(response) {
          WFSLayer.addData(response);
          WFSLayer.addTo(baseTree);
        }
    });

      var buildingsowsrootUrl = 'http://localhost:8080/geoserver/colleferro.territorio/ows';

     var buildingsdefaultParametersdue = {
         service: 'WFS',
         version: '2.0',
         request: 'GetFeature',
        typeName: 'colleferro.territorio:edifici_colleferro',
        outputFormat: 'text/javascript',
        format_options: 'callback:getJson',
        SrsName: 'EPSG:4326'
     };

    var buildingsparameters = L.Util.extend(BuildingsdefaultParameters);
    var buildingsURL = buildingsowsrootUrl + L.Util.getParamString(buildingsparameters);

    var WFSbuildingsLayer = L.geoJson(null, {
        style: function(feature) {
            return {
           stroke: true,
            fillColor: '#B04173',
            fillOpacity: 2,
            color: '#000000',
            weight: 1,
          };
        }
     });

     var ajax = $.ajax({
         url: buildingsURL,
        dataType: 'jsonp',
        jsonpCallback: 'getJson',
       success: function(response) {
        WFSbuildingsLayer.addData(response);
        WFSbuildingsLayer.addTo();
        }
    });

    //opzioni mappa

       var map = L.map('map', {
          renderer: L.canvas(),
              layers: [osm],
             center: center,
              zoom: 13
       });

      L.control.scale().addTo(map);

       var baseTree = {
        label: 'BaseLayers',
        noShow: true,
        children: [
            {
                label: 'OpenStreeMap',

                children: [
                    {label: 'OpenStreetMap Standard', layer: osm},
                ]
            },
            {
                label: 'Stadia Basemaps',
                children: [
                    {label: 'Stadia Alidade Smooth Dark', layer: stadiadark},
                ]
            },
            {
                label: 'Thunder',
                children: [
                    {label: 'Cycle', layer: cycle},
                    {label: 'Transport', layer: transport},
                ]

            },
            {
                         label: 'Bing',
                         children: [
                        {label: 'Bing Satellite', layer: binglayer}
                         ]
                     },
                ]
           };

           var overlaysTree = {
            label: 'City',
            selectAllCheckbox: 'Un/select all',
            children: [
     {

            label: 'Territorio',
            selectAllCheckbox: true,
            children: [
            {
                label: 'Test',
                selectAllCheckbox: true,
                children: [
                    { label: 'Boundaries', layer: boundaries },
                    { label:'Streets', layer: WFSLayer},
                    {label: 'Buildings', layer: buildingsWFSLayer}

                ]
            },
            /* ... */

          ]
        }
      ]
     }

     var lay = L.control.layers.tree(baseTree, overlaysTree, {
          namedToggle: true,
          selectorBack: false,
          closedSymbol: '&#8862; &#x1f5c0;',
         openedSymbol: '&#8863; &#x1f5c1;',
         collapseAll: 'Collapse all',
         expandAll: 'Expand all',
         collapsed: false,
      });

     lay.addTo(map).collapseTree().expandSelected().collapseTree(true);
     L.DomEvent.on(L.DomUtil.get('onlysel'), 'click', function() {
     lay.collapseTree(true).expandSelected(true);

    });

    </script>
    </body>

    </html>`

**Comment by jjimenezshaw** [comment]
@Matt914 first some comments:
- Please, indent your code. It is illegible. There are tools that do it automatically.
- If you define any layer in `localhost`, nobody else will be able to test. So only code inspection is possible.

Related to your problem, `WFSLayer` is loaded asynchronously. So it is not there when you create the tree. You could:
a) wait somehow until you have it to create the tree.
b) embed it in a different layer (maybe a `layerGroup`), and insert this container in the tree. When your json is loaded, insert it in the container. I am not sure if this will work.

**Comment by Matt914** [comment]
Hi, I edited my first post and added everything I've done so far. I edited with a non-async WFS request and was able to load the layer and defined my layers. What can I do to perform two or more ajax calls? If I understand correctly the problem is with the Callback request because it was already used with the first WFS.
I hope everything is a bit more readable now.
